File: src/dedupe.ts
import { createHash } from "node:crypto";

/**
 * Sender-scoped dedupe key. SHA256 hex of the parts joined by ":".
 *
 * Convention from CLAUDE.md + ARCHITECTURE.md §6.6:
 *   computeDedupeKey(["comms-adviser", staffId, channel, threadRef])
 * The first part should be the sender agent id so two agents can't
 * collide on the same natural key.
 */
export function computeDedupeKey(parts: readonly string[]): string {
  if (parts.length === 0) {
    throw new Error("computeDedupeKey: parts must be non-empty");
  }
  const h = createHash("sha256");
  h.update(parts.join(":"));
  return h.digest("hex");
}

File: python/ai_ops_control_plane/dedupe.py
"""Dedupe key — mirror src/dedupe.ts. Same SHA256 hex output for the same parts."""

from __future__ import annotations

import hashlib


def compute_dedupe_key(parts: list[str]) -> str:
    """Sender-scoped dedupe key. SHA256 hex of the parts joined by ':'.

    Convention from CLAUDE.md + ARCHITECTURE.md §6.6:
        compute_dedupe_key(["comms-adviser", staff_id, channel, thread_ref])
    The first part should be the sender agent id so two agents can't
    collide on the same natural key.
    """
    if not parts:
        raise ValueError("compute_dedupe_key: parts must be non-empty")
    return hashlib.sha256(":".join(parts).encode("utf-8")).hexdigest()
